### ISSUE TITLE
pdfDefaultOptions.textLayerMode is ignored/overridden

### ISSUE THREAD

**Issue by hdensity** [stephanrauh/ngx-extended-pdf-viewer#694]
**Describe the bug**
Setting textLayerMode to 2 (ENHANCED) in pdfDefaultOptions is ignored by NGX Extended PDF Viewer.

**Version info**
- ngx-extended-pdf-viewer: 8.2
- PDFJS: 2.7 & 2.8

**Desktop:**
 - all browsers

**To Reproduce**
- Create an Angular project with NGX Extended PDF Viewer installed and configured
- Change app.comonent.ts to the following:

```typescript
@Component({
  selector: 'app-component',
  template: '<ngx-extended-pdf-viewer [src]=pdfSrc [textLayer]=true (pageRendered)=pageRendered($event) (textLayerRendered)=textLayerRendered($event)></ngx-extended-pdf-viewer>',
})
class AppComponent {

  pdfSrc = 'assets/sample.pdf';

  constructor() {
    pdfDefaultOptions.textLayerMode = 2;
  }

  public pageRendered(event: PageRenderedEvent): void {
    console.log('text layer mode', event.source['textLayerMode']);
  }

  public textLayerRendered($event: TextLayerRenderedEvent): void {
    console.log('should use enhanced text selection', event.source['enhanceTextSelection']);
  }
}
```

The logs will show ```1``` and ```false``` respectively, instead of the expected ```2``` and ```true```.

**Workaround**
The private method ```activateTextlayerIfNecessary``` overrides the value of textLayerMode. As a workaround, add the following to your component:

```typescript
  @ViewChild(NgxExtendedPdfViewerComponent)
  set pdfViewer(pdfViewerComponent: NgxExtendedPdfViewerComponent) {
    if (pdfViewerComponent) {
      const originalFunction = pdfViewerComponent['activateTextlayerIfNecessary'].bind(pdfViewerComponent);
      Object.defineProperty(
        pdfViewerComponent,
        'activateTextlayerIfNecessary',
        {
          value: (options) => {
            originalFunction(options);
            if (options) {
              options.set('textLayerMode', 2);
            }
          }
        }
      );
    }
  }
```

**Comment by stephanrauh** [comment]
OK, this makes sense. Just a question: what is the enhanced text layer?

**Comment by hdensity** [comment]
Thank you for getting back to me.

The enhanced text layer fixes an issue with text selection. On normal textLayerMode = 1, while selecting text, if your mouse does not hover over a span element of the text layer, everything on the page below the start of your selection is selected. Or in other words, it selects everything from the start of the selection to the first/last span of the next/previous page respectively. In some cases, when for example the spacing between lines is big enough, this can lead to a flickering text selection while moving the mouse.
With enhanced textLayerMode = 2, this behaviour is fixed. The devs of pdf.js implemented what they called div expansion, basically ensuring the mouse never "touches" the page/canvas layer (except for top or bottom page borders, but hey, it's a huge improvement). The only downside is a small hit on rendering performance, due to the added calculation steps, but I couldn't feel a difference in my usage.

For reference: https://github.com/mozilla/pdf.js/pull/7539

**Comment by stephanrauh** [comment]
Just a short sign-of-life: I've decided to examine the enhance text layer mode. If the feature works as you've described it, I wonder why we still use `textLayerMode=1` as default.

**Comment by stephanrauh** [comment]
The feature has landed with version 8.3.0. I've also decided to make the enhanced text layer the default.

Best regards,
Stephan

**Comment by hdensity** [comment]
Thank you very much Stephan, much appreciated.

Best regards,
Sam
